File: packages/rules/ts/src/engine.ts
import bundleJson from "../../dist/rules.json";
import { evaluateNode, UNKNOWN } from "./evaluate";
import { FIELDS } from "./fields";
import type {
  MatchResult,
  MatchRun,
  NextQuestion,
  Profile,
  Reason,
  RuleBundle,
  Scheme,
  Verdict,
} from "./types";

export const bundle = bundleJson as unknown as RuleBundle;
export const ENGINE_VERSION = bundle.engine_version;
export const RULES_DIGEST = bundle.rules_digest;

/**
 * Whether a speaker of `language` has reviewed the copy. Compiled into the bundle by
 * the Python side so the browser and the API cannot disagree about it.
 */
export function translationStatus(language: string): string {
  return bundle.translation_status?.[language] ?? "fallback";
}

const NO_CEILING = Number.POSITIVE_INFINITY;

const VERDICT_ORDER: Record<Verdict, number> = {
  ELIGIBLE: 0,
  LIKELY_ELIGIBLE: 1,
  NEED_MORE_INFO: 2,
  INELIGIBLE: 3,
};

function message(mapping: Record<string, string>, language: string): string {
  return mapping[language] || mapping.en || "";
}

/** Match Python's `round(value, places)` for the magnitudes this engine produces. */
function roundTo(value: number, places: number): number {
  const factor = 10 ** places;
  return Math.round(value * factor) / factor;
}

export function indicativeAmount(scheme: Scheme, profile: Profile): number | null {
  const cost = profile.project_cost;
  if (cost === null || cost === undefined) return null;

  const pct = scheme.limits.max_funding_pct;
  const funded = pct !== null ? (cost as number) * (pct / 100) : (cost as number);
  // The loan cap, not the project-cost band. See the Python docstring.
  const cap = scheme.limits.max_loan_amount;
  return roundTo(cap === null ? funded : Math.min(funded, cap), 2);
}

export function evaluateScheme(
  scheme: Scheme,
  profile: Profile,
  language = "en",
): MatchResult {
  const blocked: Reason[] = [];
  const matched: Reason[] = [];
  const warnings: Reason[] = [];
  const missing = new Set<string>();
  let redirect: string | null = null;

  const hardRules = scheme.rules.filter((r) => r.severity === "HARD_BLOCK");
  let decidedHard = 0;

  for (const rule of scheme.rules) {
    const outcome = evaluateNode(rule.when, profile);

    if (outcome === UNKNOWN) {
      if (rule.severity === "HARD_BLOCK") {
        for (const field of rule.fields) {
          const value = profile[field];
          if (value === null || value === undefined) missing.add(field);
        }
      }
      continue;
    }

    if (rule.severity === "HARD_BLOCK") decidedHard += 1;

    if (outcome === true) {
      const reason: Reason = {
        rule_id: rule.id,
        message: message(rule.message_i18n, language),
        severity: rule.severity,
      };
      if (rule.severity === "HARD_BLOCK") blocked.push(reason);
      else warnings.push(reason);
      if (redirect === null && rule.suggest_instead) redirect = rule.suggest_instead;
    } else if (Object.keys(rule.satisfied_i18n).length > 0) {
      matched.push({
        rule_id: rule.id,
        message: message(rule.satisfied_i18n, language),
        severity: rule.severity,
      });
    }
  }

  let verdict: Verdict;
  let confidence: number;
  if (blocked.length > 0) {
    verdict = "INELIGIBLE";
    confidence = 1.0;
  } else if (missing.size > 0) {
    verdict = "NEED_MORE_INFO";
    confidence = hardRules.length ? roundTo(decidedHard / hardRules.length, 4) : 0.0;
  } else if (warnings.length > 0) {
    verdict = "LIKELY_ELIGIBLE";
    confidence = 1.0;
  } else {
    verdict = "ELIGIBLE";
    confidence = 1.0;
  }

  const band =
    scheme.limits.interest_rate_min !== null && scheme.limits.interest_rate_max !== null
      ? ([scheme.limits.interest_rate_min, scheme.limits.interest_rate_max] as [number, number])
      : null;

  return {
    scheme_code: scheme.code,
    official_name: scheme.official_name,
    family: scheme.family,
    verdict,
    confidence,
    matched_because: matched,
    blocked_because: blocked,
    warnings,
    missing_fields: [...missing].sort(),
    indicative_amount: verdict === "INELIGIBLE" ? null : indicativeAmount(scheme, profile),
    indicative_interest_band: band,
    max_funding_pct: scheme.limits.max_funding_pct,
    redirect_suggestion: redirect,
    needs_verification: scheme.provenance.needs_verification,
    provenance: scheme.provenance,
    rank: null,
    translation_status: translationStatus(language),
  };
}

function rankKey(scheme: Scheme, profile: Profile): [number, number, number] {
  const cost = profile.project_cost;
  const ceiling = scheme.limits.max_project_cost;

  let headroom = NO_CEILING;
  if (cost !== null && cost !== undefined && ceiling !== null) {
    const slack = ceiling - (cost as number);
    headroom = slack < 0 ? NO_CEILING : slack;
  }

  const interest = scheme.limits.interest_rate_min ?? NO_CEILING;
  const funding = -(scheme.limits.max_funding_pct ?? 0);
  return [headroom, interest, funding];
}

export function evaluate(profile: Profile, language = "en"): MatchResult[] {
  const byCode = new Map(bundle.schemes.map((s) => [s.code, s]));
  const results = bundle.schemes.map((s) => evaluateScheme(s, profile, language));

  results.sort((a, b) => {
    const va = VERDICT_ORDER[a.verdict];
    const vb = VERDICT_ORDER[b.verdict];
    if (va !== vb) return va - vb;

    const ka = rankKey(byCode.get(a.scheme_code)!, profile);
    const kb = rankKey(byCode.get(b.scheme_code)!, profile);
    for (let i = 0; i < ka.length; i += 1) {
      if (ka[i] !== kb[i]) return ka[i] < kb[i] ? -1 : 1;
    }
    return a.scheme_code < b.scheme_code ? -1 : a.scheme_code > b.scheme_code ? 1 : 0;
  });

  return results.map((r, i) => ({ ...r, rank: i + 1 }));
}

export function fieldImpact(
  profile: Profile,
): Record<string, { rules: Set<string>; schemes: Set<string> }> {
  const impact: Record<string, { rules: Set<string>; schemes: Set<string> }> = {};

  for (const scheme of bundle.schemes) {
    const hardRules = scheme.rules.filter((r) => r.severity === "HARD_BLOCK");
    const outcomes = hardRules.map((rule) => [rule, evaluateNode(rule.when, profile)] as const);

    // A scheme with a definite block is settled; asking more about it changes nothing.
    if (outcomes.some(([, outcome]) => outcome === true)) continue;

    for (const [rule, outcome] of outcomes) {
      if (outcome !== UNKNOWN) continue;
      for (const field of rule.fields) {
        const value = profile[field];
        if (value !== null && value !== undefined) continue;
        impact[field] ??= { rules: new Set(), schemes: new Set() };
        impact[field].rules.add(rule.id);
        impact[field].schemes.add(scheme.code);
      }
    }
  }
  return impact;
}

/**
 * `exclude` names fields already put to the citizen. Asking the same question every
 * turn because they keep not answering it makes no progress and reads as broken.
 */
export function nextBestQuestion(
  profile: Profile,
  exclude?: Iterable<string>,
): NextQuestion | null {
  const impact = fieldImpact(profile);
  const skip = new Set(exclude ?? []);
  const remaining = Object.keys(impact).filter((f) => !skip.has(f));
  const names = remaining.length > 0 ? remaining : Object.keys(impact);
  if (names.length === 0) return null;

  names.sort((a, b) => {
    const byCount = impact[b].rules.size - impact[a].rules.size;
    if (byCount !== 0) return byCount;
    const byPriority = FIELDS[a].priority - FIELDS[b].priority;
    if (byPriority !== 0) return byPriority;
    return a < b ? -1 : 1;
  });

  const name = names[0];
  const field = FIELDS[name];
  return {
    field: name,
    question_i18n: { ...field.question_i18n },
    kind: field.kind,
    choices: field.choices ?? null,
    resolves_rules: [...impact[name].rules].sort(),
    resolves_schemes: [...impact[name].schemes].sort(),
  };
}

export function run(profile: Profile, language = "en"): MatchRun {
  const snapshot: Record<string, unknown> = {};
  for (const key of Object.keys(profile).sort()) snapshot[key] = profile[key];

  return {
    engine_version: ENGINE_VERSION,
    rules_digest: RULES_DIGEST,
    input_snapshot: snapshot,
    results: evaluate(profile, language),
    next_question: nextBestQuestion(profile),
  };
}

File: packages/rules/src/setu_rules/engine.py
"""The deterministic eligibility engine.

No language model participates in any decision made here. Given the same profile and
the same rule files, this module returns byte-identical output every time — that is
what makes a verdict replayable from `match_runs` months later.

Verdict logic, in order:

  1. Any HARD_BLOCK rule that is definitively true  -> INELIGIBLE
  2. Otherwise, any HARD_BLOCK rule that is UNKNOWN -> NEED_MORE_INFO
  3. Otherwise, any SOFT_WARN rule that is true     -> LIKELY_ELIGIBLE
  4. Otherwise                                      -> ELIGIBLE

Step 2 is the important one. A rule is UNKNOWN when it references a field the citizen
has not given us. We report that as "we need one more fact", never as "you do not
qualify".
"""

from __future__ import annotations

from dataclasses import replace
from typing import Any

from setu_rules.loader import load_schemes, rules_digest
from setu_rules.models import (
    MatchResult,
    MatchRun,
    Reason,
    Scheme,
    Severity,
    Verdict,
)
from setu_rules.profile import validate_profile
from setu_rules.questions import next_best_question
from setu_rules.safe_eval import UNKNOWN, evaluate_node
from setu_rules.translations import translation_status
from setu_rules.version import ENGINE_VERSION

# Sentinel used when a scheme states no ceiling, so it sorts after any scheme that
# does state one. Larger than the largest figure in the problem statement.
_NO_CEILING = float("inf")


def _message(mapping: dict[str, str], language: str) -> str:
    """Locale lookup with an explicit English fallback.

    Missing translations fall back rather than raising: a citizen reading Marathi still
    gets a correct reason, in English, instead of an error page.
    """
    return mapping.get(language) or mapping.get("en") or ""


def evaluate_scheme(scheme: Scheme, profile: dict[str, Any], language: str = "en") -> MatchResult:
    """Evaluate one scheme against one profile."""
    blocked: list[Reason] = []
    matched: list[Reason] = []
    warnings: list[Reason] = []
    missing: set[str] = set()
    redirect: str | None = None

    hard_rules = [r for r in scheme.rules if r.severity is Severity.HARD_BLOCK]
    decided_hard = 0

    for rule in scheme.rules:
        outcome = evaluate_node(rule.when, profile)

        if outcome is UNKNOWN:
            if rule.severity is Severity.HARD_BLOCK:
                # Only the fields actually absent are worth asking about.
                missing.update(f for f in rule.fields if profile.get(f) is None)
            continue

        if rule.severity is Severity.HARD_BLOCK:
            decided_hard += 1

        if outcome is True:
            reason = Reason(rule.id, _message(rule.message_i18n, language), rule.severity)
            if rule.severity is Severity.HARD_BLOCK:
                blocked.append(reason)
                # First redirect wins; rules are evaluated in file order, which is
                # authored most-specific-first.
                if redirect is None and rule.suggest_instead:
                    redirect = rule.suggest_instead
            else:
                warnings.append(reason)
                if redirect is None and rule.suggest_instead:
                    redirect = rule.suggest_instead
        elif rule.satisfied_i18n:
            matched.append(
                Reason(rule.id, _message(rule.satisfied_i18n, language), rule.severity)
            )

    if blocked:
        verdict = Verdict.INELIGIBLE
        confidence = 1.0
    elif missing:
        verdict = Verdict.NEED_MORE_INFO
        confidence = round(decided_hard / len(hard_rules), 4) if hard_rules else 0.0
    elif warnings:
        verdict = Verdict.LIKELY_ELIGIBLE
        confidence = 1.0
    else:
        verdict = Verdict.ELIGIBLE
        confidence = 1.0

    return MatchResult(
        scheme_code=scheme.code,
        official_name=scheme.official_name,
        family=scheme.family,
        verdict=verdict,
        confidence=confidence,
        matched_because=tuple(matched),
        blocked_because=tuple(blocked),
        warnings=tuple(warnings),
        missing_fields=tuple(sorted(missing)),
        # No figure for a scheme the citizen cannot have — an amount on a blocked card
        # is an anchor they will remember and we would be inventing it.
        indicative_amount=(
            None if verdict is Verdict.INELIGIBLE else indicative_amount(scheme, profile)
        ),
        indicative_interest_band=(
            (scheme.limits.interest_rate_min, scheme.limits.interest_rate_max)
            if scheme.limits.interest_rate_min is not None
            and scheme.limits.interest_rate_max is not None
            else None
        ),
        max_funding_pct=scheme.limits.max_funding_pct,
        redirect_suggestion=redirect,
        needs_verification=scheme.provenance.needs_verification,
        provenance=scheme.provenance,
        translation_status=translation_status(language),
    )


def indicative_amount(scheme: Scheme, profile: dict[str, Any]) -> float | None:
    """The smaller of the scheme's LOAN cap and the funded share of the project cost.

    The loan cap, not the project-cost band: NSFDC will fund a Rs 1,40,000 micro-finance
    unit but lend at most Rs 1,25,000 against it. Using the band here would tell the
    citizen they can borrow Rs 1,26,000, which is false.

    Returns None when the project cost is unknown, rather than showing a figure the
    citizen might anchor on.
    """
    cost = profile.get("project_cost")
    if cost is None:
        return None

    pct = scheme.limits.max_funding_pct
    funded = cost * (pct / 100.0) if pct is not None else float(cost)

    cap = scheme.limits.max_loan_amount
    amount = funded if cap is None else min(funded, float(cap))
    return round(amount, 2)


def _rank_key(result: MatchResult, scheme: Scheme, profile: dict[str, Any]) -> tuple:
    """Pure ranking key: tightest fit, then cheapest, then most generous.

    The trailing scheme_code guarantees a total order, so ranking never depends on
    dict or filesystem iteration order.
    """
    cost = profile.get("project_cost")
    ceiling = scheme.limits.max_project_cost

    if cost is None or ceiling is None:
        headroom = _NO_CEILING
    else:
        headroom = float(ceiling) - float(cost)
        if headroom < 0:
            headroom = _NO_CEILING  # cannot actually fund the ask

    interest = scheme.limits.interest_rate_min
    interest_key = interest if interest is not None else _NO_CEILING

    pct = scheme.limits.max_funding_pct
    funding_key = -(pct if pct is not None else 0.0)

    return (headroom, interest_key, funding_key, result.scheme_code)


_VERDICT_ORDER = {
    Verdict.ELIGIBLE: 0,
    Verdict.LIKELY_ELIGIBLE: 1,
    Verdict.NEED_MORE_INFO: 2,
    Verdict.INELIGIBLE: 3,
}


def evaluate(
    profile: dict[str, Any],
    language: str = "en",
    schemes: tuple[Scheme, ...] | None = None,
) -> list[MatchResult]:
    """Evaluate every scheme and return them in deterministic ranked order.

    Ineligible schemes are returned too, ranked last. The UI shows them greyed out with
    the exact blocking reason — transparency beats hiding.
    """
    validate_profile(profile)
    catalogue = schemes if schemes is not None else load_schemes()
    by_code = {s.code: s for s in catalogue}

    results = [evaluate_scheme(s, profile, language) for s in catalogue]
    results.sort(
        key=lambda r: (_VERDICT_ORDER[r.verdict], *_rank_key(r, by_code[r.scheme_code], profile))
    )

    return [replace(r, rank=i) for i, r in enumerate(results, start=1)]


def run(
    profile: dict[str, Any],
    language: str = "en",
    schemes: tuple[Scheme, ...] | None = None,
) -> MatchRun:
    """Evaluate and wrap the result in the record persisted to `match_runs`."""
    catalogue = schemes if schemes is not None else load_schemes()
    results = evaluate(profile, language, catalogue)
    return MatchRun(
        engine_version=ENGINE_VERSION,
        rules_digest=rules_digest(),
        input_snapshot=dict(sorted(profile.items())),
        results=tuple(results),
        next_question=next_best_question(profile, catalogue),
    )
